Fix tuple comparison translator generating invalid `IN` comparisons.

ndm2 opened the pull request:
Before this change, the tuple comparison translator would rewrite the conditions of the subquery used in the provided test case to something that would generate SQL like this:

```sql
WHERE (
    ArticlesAlias.author_id = :c0 AND
    Articles.id IN ArticlesAlias.id AND
    Articles.author_id IN ArticlesAlias.author_id
)
```

Those `IN` comparisons are invalid. Before #15777 the field on the right hand side would have been wrapped in parentheses, eg

```sql
Articles.id IN (ArticlesAlias.id) AND
Articles.author_id IN (ArticlesAlias.author_id)
```

so it worked by accident so to speak.

With this change, the comparisons will now use `=` instead of `IN`, eg

```sql
Articles.id = ArticlesAlias.id AND
Articles.author_id = ArticlesAlias.author_id
```

ndm2 committed:
Fix tuple comparison translator generating invalid `IN` comparisons.

`IdentifierExpression` values for `IN` comparisons are not being put
into parentheses anymore, resulting in invalid SQL like
`Table.column_a IN Table.column_b` when unconditionally reusing the
expression's operator.

othercorey commented on src/Database/Driver/TupleComparisonTranslatorTrait.php:
```diff
@@ -56,13 +56,16 @@ protected function _transformTupleComparison(TupleComparison $expression, Query
+            $op = $expression->getOperator();
+            if (strtolower($op) === 'in') {
+                $op = '=';
+            }
```
Is this supposed to support `NOT IN` as well? It doesn't seem like it would.

othercorey replied:
I still think we should reject unsupported operators.

markstory replied:
Won't invalid operators also make invalid SQL which will fail when run?

othercorey replied:
Technically, it will now that identifier expressions are not wrapped in ().

It seems like we're suggesting it supports more than it does though.

ndm2 replied:
`NOT IN` isn't supported currently at all, neither for subqueries nor for arrays of values. I thought this was "intentional" 🤔 The docs also don't specifically suggest that it would work, they only mention `IN`.

othercorey replied:
It's not a big issue. Just thinking ahead now that you found this hidden issue.

ndm2 replied:
Yeah, I guess it's time to think about a possible alternative solution.

I'm also wondering whether this is actually still needed for Sqlite and subqueries. I'd have to take a closer look, but it seems with 3.26 composite keys seem to work fine with subqueries, and for an array of values it could possibly use `VALUES`. Might apply to 3.x in general.

ADmad commented on src/Database/Driver/TupleComparisonTranslatorTrait.php:
```diff
@@ -56,13 +56,16 @@ protected function _transformTupleComparison(TupleComparison $expression, Query
 
         if ($value instanceof Query) {
+            $op = $expression->getOperator();
+            if (strtolower($op) === 'in') {
```
Nitpick: Can we uppercase instead for better readability? :slightly_smiling_face:

ndm2 committed:
Accept only working operators for tuple comparison transformation.

ndm2 committed:
Add tuple comparison integration tests.

ndm2 committed:
Handle invalid tuple comparison expression values more gracefully.

Throw a meaningful exception instead of running into PHP errors or
incidental exceptions later in the compilation stage.

ndm2 committed:
Fix connection test leaving behind an open transaction.

ndm2 committed:
Fix test case method name typo.

othercorey commented on src/Database/Expression/TupleComparison.php:
```diff
@@ -69,6 +70,26 @@ public function getType(): array
+            if (
+                is_array($value) &&
+                is_array(current($value))
+            ) {
```
These can probably be collapsed to a single line.

ndm2 replied:
They could, I only do that so that in case of partial matches line coverage shows which conditions have been met.

othercorey replied:
They have to both be met for the line below to get coverage though.

ndm2 replied:
Yeah, that was just the origin story, I can't get rid of that habit anymore, and now there's multi-line conditions everywhere :)

othercorey commented:
Looking good!

ndm2 committed:
Collapse redundant multi-line conditions.
